NullPointerException: Null ipamOptions

## Description
[During getting list of tasks i get NPE]
## How to reproduce
[docker version 17.06.0-ce, API version:  1.30]

## What do you expect
[Will return a list of tasks]

## What happened instead
[NPE]

## Software:
- `docker 17.06.0-ce`: [Add the output of `docker version` here, both client and server]
- Spotify's docker-client version: [8.8.0]

## Full backtrace

```text
[Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.swarm.NetworkSpec, problem: Null ipamOptions
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@2297757e; line: 1, column: 2703] (through reference chain: java.util.ArrayList[0]->com.spotify.docker.client.messages.swarm.Task["NetworksAttachments"]->java.util.ArrayList[0]->com.spotify.docker.client.messages.swarm.NetworkAttachment["Network"]->com.spotify.docker.client.messages.swarm.Network["Spec"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:277)
	at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1441)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:476)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:495)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:276)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:228)
	at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:138)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:471)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:511)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:400)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:511)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:400)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:287)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:259)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:26)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeWithErrorWrapping(BeanDeserializer.java:511)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:400)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1191)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:287)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:259)
	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:26)
	at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:1583)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:964)
	at com.fasterxml.jackson.jaxrs.base.ProviderBase.readFrom(ProviderBase.java:815)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.invokeReadFrom(ReaderInterceptorExecutor.java:256)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$TerminalReaderInterceptor.aroundReadFrom(ReaderInterceptorExecutor.java:235)
	at org.glassfish.jersey.message.internal.ReaderInterceptorExecutor.proceed(ReaderInterceptorExecutor.java:155)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.readFrom(MessageBodyFactory.java:1085)
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:874)
	at org.glassfish.jersey.message.internal.InboundMessageContext.readEntity(InboundMessageContext.java:834)
	at org.glassfish.jersey.client.ClientResponse.readEntity(ClientResponse.java:368)
	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:871)
	... 33 more
Caused by: java.lang.NullPointerException: Null ipamOptions
	at com.spotify.docker.client.messages.swarm.AutoValue_NetworkSpec.<init>(AutoValue_NetworkSpec.java:41)
	at com.spotify.docker.client.messages.swarm.NetworkSpec.create(NetworkSpec.java:75)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.fasterxml.jackson.databind.introspect.AnnotatedMethod.call(AnnotatedMethod.java:117)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:274)
	... 74 more]
```

We have discovered a similar issue when processing following response from interface "list networks";

 ... **"IPAM":{"Driver":"","Options":null,"Config":null}** ...

This exception is raised:

_Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.spotify.docker.client.messages.Ipam, problem: null
at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@1a41af83; line: 1, column: 195] (through reference chain: java.util.ArrayList[0]->com.spotify.docker.client.messages.Network["IPAM"])_

**Solution**:
In class: com.spotify.docker.client.messages.Ipam the field Driver and Config should be Nullable:

```
  @Nullable
  @JsonProperty("Driver")
  public abstract String driver();

  @Nullable
  @JsonProperty("Config")
  public abstract ImmutableList<IpamConfig> config();
```

I hope someone could apply this minor fix? Thx!

--- a/src/main/java/com/spotify/docker/client/messages/Ipam.java
+++ b/src/main/java/com/spotify/docker/client/messages/Ipam.java
@@ -31,6 +31,8 @@ import com.google.auto.value.AutoValue;
 import com.google.common.collect.ImmutableList;
 import java.util.List;
 
+import javax.annotation.Nullable;
+
 @AutoValue
 @JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
 public abstract class Ipam {
@@ -38,6 +40,7 @@ public abstract class Ipam {
   @JsonProperty("Driver")
   public abstract String driver();
 
+  @Nullable
   @JsonProperty("Config")
   public abstract ImmutableList<IpamConfig> config();
 

--- a/src/test/java/com/spotify/docker/client/messages/IpamTest.java
+++ b/src/test/java/com/spotify/docker/client/messages/IpamTest.java
@@ -1,0 +1,54 @@
+/*-
+ * -\-\-
+ * docker-client
+ * --
+ * Copyright (C) 2016 - 2017 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.docker.client.messages;
+
+import static com.spotify.docker.FixtureUtil.fixture;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.spotify.docker.client.ObjectMapperProvider;
+
+import org.junit.Test;
+
+public class IpamTest {
+
+  private static final ObjectMapper objectMapper = ObjectMapperProvider.objectMapper();
+
+  @Test
+  public void testDeserialize() throws Exception {
+    final Ipam ipam =
+        objectMapper.readValue(fixture("fixtures/1.29/ipam.json"), Ipam.class);
+    assertThat(ipam.driver(), equalTo("default"));
+    assertThat(ipam.config(), contains(IpamConfig.create("172.17.0.0/16", null, null)));
+  }
+
+  @Test
+  public void testDeserialize_nullConfig() throws Exception {
+    final Ipam ipam =
+        objectMapper.readValue(fixture("fixtures/1.29/ipam-null-config.json"), Ipam.class);
+    assertThat(ipam.driver(), equalTo("default"));
+    assertThat(ipam.config(), is(nullValue()));
+  }
+}

--- a/src/test/resources/fixtures/1.29/ipam-null-config.json
+++ b/src/test/resources/fixtures/1.29/ipam-null-config.json
@@ -1,0 +1,3 @@
+{
+  "Driver": "default"
+}

--- a/src/test/resources/fixtures/1.29/ipam.json
+++ b/src/test/resources/fixtures/1.29/ipam.json
@@ -1,0 +1,8 @@
+{
+  "Driver": "default",
+  "Config": [
+    {
+      "Subnet": "172.17.0.0/16"
+    }
+  ]
+}

